src/detect-transform-base64.c

```diff
@@ -0,0 +1,333 @@
+            DetectTransformFromBase64DecodeRegisterTests;
+#endif
+    sigmatch_table[DETECT_TRANSFORM_FROM_BASE64].flags |= SIGMATCH_OPTIONAL_OPT;
+}
+
+static void DetectTransformFromBase64DecodeFree(DetectEngineCtx *de_ctx, void *ptr)
+{
+    SCTransformBase64Free(ptr);
+}
+
+static SCDetectTransformFromBase64Data *DetectTransformFromBase64DecodeParse(const char *str)
+{
+    SCDetectTransformFromBase64Data *tbd = SCTransformBase64Parse(str);
+    if (tbd == NULL) {
+        SCLogError("invalid transform_base64 values");
+    }
+    return tbd;
+}
+
+/**
+ *  \internal
+ *  \brief Base64 decode the input buffer
+ *  \param det_ctx detection engine ctx
+ *  \param s signature
+ *  \param opts_str transform options, if any
+ *  \retval 0 No decode
+ *  \retval >0 Decoded byte count
+ */
+static int DetectTransformFromBase64DecodeSetup(
+        DetectEngineCtx *de_ctx, Signature *s, const char *opts_str)
+{
+    int r = -1;
+
+    SCEnter();
+
+    SCDetectTransformFromBase64Data *b64d = DetectTransformFromBase64DecodeParse(opts_str);
+    if (b64d == NULL)
+        SCReturnInt(r);
+
+    if (b64d->flags & DETECT_TRANSFORM_BASE64_FLAG_OFFSET_VAR) {
+        SCLogError("offset value must be a value, not a variable name");
+        goto exit_path;
+    }
+
+    if (b64d->flags & DETECT_TRANSFORM_BASE64_FLAG_NBYTES_VAR) {
+        SCLogError("byte value must be a value, not a variable name");
+        goto exit_path;
+    }
+
+    r = DetectSignatureAddTransform(s, DETECT_TRANSFORM_FROM_BASE64, b64d);
+
+exit_path:
+    if (r != 0)
+        DetectTransformFromBase64DecodeFree(de_ctx, b64d);
+    SCReturnInt(r);
+}
+
+static void TransformFromBase64Decode(InspectionBuffer *buffer, void *options)
+{
+    SCDetectTransformFromBase64Data *b64d = options;
+    const uint8_t *input = buffer->inspect;
+    const uint32_t input_len = buffer->inspect_len;
+    /*
+     * The output buffer is larger than needed. On average,
+     * a base64 encoded string is 75% of the unencoded
+     * string.
+     */
+    uint8_t output[input_len];
+    uint32_t decode_length = input_len;
+
+    DetectBase64Mode mode = b64d->mode;
+    uint32_t offset = b64d->offset;
+    uint32_t nbytes = b64d->nbytes;
+
+    if (offset) {
+        if (offset > input_len) {
+            SCLogDebug("offset %d exceeds length %d; returning", offset, input_len);
+            return;
+        }
+        input += offset;
+        decode_length -= offset;
+    }
+
+    if (nbytes) {
+        if (nbytes > decode_length) {
+            SCLogDebug("byte count %d plus offset %d exceeds length %d; returning", nbytes, offset,
+                    input_len);
+            return;
+        }
+        decode_length = nbytes;
+    }
+
+    // PrintRawDataFp(stdout, input, input_len);
+    uint32_t decoded_length;
+    uint32_t consumed;
+    Base64Ecode code =
+            DecodeBase64(output, input_len, input, decode_length, &consumed, &decoded_length, mode);
+    if (code == BASE64_ECODE_OK) {
```

Comment:
What should we do if this was not base64 ? `InspectionBufferCopy(buffer, output, 0)` ?

Reply:
That's a good question.

This transform follows the lead of `url_decode` :
```
    changed = BufferUrlDecode(input, input_len, output, &output_size);

    if (changed) {
        InspectionBufferCopy(buffer, output, output_size);
    }
 ```

Reply:
Thoughts @victorjulien ?

I would expect that a rule cannot match if the content is not base64

Reply:
Seems like a good question to put in front of rule writers.

Reply:
I've posed the question to #rule-lang-dev on discord.

Reply:
Brandon Murphy indicated that
- Decode errors should not match. Since the buffer is unchanged on an error, a match with decoded data won't occur
- Partial decodes should return the partial decoded buffer. I'll make an update to handle the partial decode case.
- He asked if `relative` should be added. For a transform, `relative` wouldn't have the same meaning as it would with `content` or other keywords.

Reply:
> Brandon Murphy indicated that
> 
> * Decode errors should not match. Since the buffer is unchanged on an error, a match with decoded data won't occur

This looks wrong to me :
`?toto` would be a decode error but still match on `toto` content as `dG90bw==` would

Reply:
> This looks wrong to me : `?toto` would be a decode error but still match on `toto` content as `dG90bw==` would

That should be taken care of by the rule writer with an offset kind of option, I think.
So, for `?dG90bw==%YmxhaA==`, one should be able to write and match on
1. `from_base64: offset: 1; content: toto;` to match on `toto` and,
2. `from_base64: offset: 6; content: blah;` to match on `blah` etc.

wdyt @catenacyber ?

Note @jlucovsky: Partial decoding is possible under RFC 4648 as long as it is not strict mode. If you see a different behavior then it may be a bug.
So, the idea is:
|RFC| What it does | Used by |
|-----|-------------------|---------------|
|RFC 2045| Skip over any invalid chars and decode the resulting base64 encoded string | Mime parser | 
|RFC 4648| Decode until an invalid char is encountered | base64_decode |
|Strict mode (also uses RFC 4648)| If an invalid char is encountered, reset decoded bytes to 0 and return an error| Datasets|